engine/filesync/README.md

```diff
@@ -0,0 +1,13 @@
+This package is mostly extracted from Buildkit v0.10.5's [`session/filesync`][src].
+
+The following changes are made:
```

Comment:
Totally fine with copying this code for now. Not a blocker, but in the medium-to-long term I do think it's worth seeing if we could modify it upstream to support our use case in a way that doesn't break theirs. E.g. maybe we could merge a change upstream that adds a method that lets you plug in more localdirs after the session has started.

If that's acceptable upstream, it would be nice and slightly reduce maintenance burden for us. If it's not, it's no biggie though, this isn't exactly a ton of code to maintain here.

Reply:
Yeah makes sense!

One way to do this would be changing the static list of directories to an interface like `Lookup(name string) (SyncedDir, bool)`. Then you'd just add the attachable to the session yourself and skip the `LocalDirs` config. Should be backwards-compatible.

We could inject one that just always returns a valid directory for the requested path, and then for more risky areas like the Playground we could have it just always return `false`.

Going back upstream _might_ close the door on fancier ways to sync - for example, with a remote engine it seems like we'll need a path like this for filesyncing paths from the client host to the engine host:

```mermaid
graph LR;
  C[Dagger Client]-->A[Engine API];
  subgraph Engine Host
  A-->B[Buildkit];
  end
```

I'll see if there's a way to still support that via upstream changes, but it might be more challenging since all the internal code assumes a `fs.FS`-like interface. I guess there could be an implementation of that that reads from a one-shot `tar` stream? :thinking:

Reply:
Yeah `Lookup(name string)` makes sense to me! 

> but it might be more challenging since all the internal code assumes a fs.FS-like interface. I guess there could be an implementation of that that reads from a one-shot tar stream?

Yeah if we could eventually modify upstream to support essentially `Lookup(name string) (fsutil.FS, error)` that seems like a reasonable possible route (having spent a total of several seconds considering it).

Reply:
Also, this has apparently been discussed in slightly different form before: https://github.com/moby/buildkit/discussions/2744#discussioncomment-2465469

Reply:
Opened https://github.com/moby/buildkit/pull/3228

I just stuck with a simple lookup that returns a `SyncedDir` for now. The more I look at how the `fsutil.FS` part works the less sure I am that we'd be able to use it for the remote engine change. It seems to kick off parallel reads and stuff. I think we might actually just need some kind of lower-level proxy.